File: platforms/web/alethion/greatfunctions/great_fn_client.js
export class GreatFnId {
  constructor(city, district, block, lane) {
    this.city = city;
    this.district = district;
    this.block = block;
    this.lane = lane;
  }
  toString() {
    return `${this.city}-${this.district}-${this.block}-${this.lane}`;
  }
}

export class GreatFnCall {
  constructor(target, cityZone, epoch, payload = {}, priority = 0) {
    this.target = target;
    this.cityZone = cityZone;
    this.epoch = epoch;
    this.payload = payload;
    this.priority = priority;
  }
}

export class GreatFnClient {
  constructor(transport) {
    this.transport = transport;
  }

  route(call) {
    const body = {
      target: call.target.toString(),
      zone: call.cityZone,
      epoch: call.epoch,
      payload: call.payload,
      priority: call.priority,
    };
    return this.transport.post("/great_fn/route", body);
  }

  tick() {
    return this.transport.post("/great_fn/tick", {});
  }
}

File: platforms/android/alethion/greatfunctions/src/main/java/city/alethion/greatfunctions/GreatFnClient.kt
package city.alethion.greatfunctions

data class GreatFnId(
    val city: Int,
    val district: Int,
    val block: Int,
    val lane: Int
) {
    override fun toString(): String = "$city-$district-$block-$lane"
}

data class GreatFnCall(
    val target: GreatFnId,
    val cityZone: Int,
    val epoch: Long,
    val payloadScalar: Long? = null,
    val payloadVector: IntArray? = null,
    val payloadGrid: ShortArray? = null,
    val priority: Int = 0
)

interface GreatFnGateway {
    fun route(call: GreatFnCall)
    fun tick()
}

class InMemoryGreatFnGateway : GreatFnGateway {
    private val queue: ArrayDeque<GreatFnCall> = ArrayDeque()

    override fun route(call: GreatFnCall) {
        queue.addLast(call)
    }

    override fun tick() {
        var processed = 0
        while (queue.isNotEmpty() && processed < 64) {
            queue.removeFirst()
            processed++
        }
    }

    fun queueSize(): Int = queue.size
}
